PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azuretools/ijidea/ui/DeviceLoginWindow.java

```diff
@@ -0,0 +1,139 @@
+ * SOFTWARE.
+ */
+
+package com.microsoft.azuretools.ijidea.ui;
+
+import com.intellij.openapi.application.ApplicationManager;
+import com.intellij.openapi.application.ModalityState;
+import com.microsoft.aad.adal4j.AdalErrorCode;
+import com.microsoft.aad.adal4j.AuthenticationCallback;
+import com.microsoft.aad.adal4j.AuthenticationContext;
+import com.microsoft.aad.adal4j.AuthenticationException;
+import com.microsoft.aad.adal4j.AuthenticationResult;
+import com.microsoft.aad.adal4j.DeviceCode;
+import com.microsoft.intellij.ui.components.AzureDialogWrapper;
+import org.jetbrains.annotations.Nullable;
+import java.awt.Desktop;
+import java.awt.Window;
+import java.awt.event.WindowEvent;
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.swing.Action;
+import javax.swing.JComponent;
+import javax.swing.JEditorPane;
+import javax.swing.JPanel;
+import javax.swing.event.HyperlinkEvent;
+
+public class DeviceLoginWindow extends AzureDialogWrapper {
+    private JPanel jPanel;
+    private JEditorPane editorPanel;
+    private AuthenticationResult authenticationResult = null;
+    private Future<?> authExecutor;
+    private static final String TITLE = "Azure Device Login";
+
+    public AuthenticationResult getAuthenticationResult() {
+        return authenticationResult;
+    }
+
+    public DeviceLoginWindow(final AuthenticationContext ctx, final DeviceCode deviceCode,
+                             final AuthenticationCallback<AuthenticationResult> callBack) {
+        super(null, false, IdeModalityType.PROJECT);
+        setModal(true);
+        setTitle(TITLE);
+        editorPanel.setText(createHtmlFormatMessage(deviceCode.getMessage()));
+        editorPanel.addHyperlinkListener(e -> {
+            if (e.getEventType() == HyperlinkEvent.EventType.ACTIVATED) {
+                try {
+                    Desktop.getDesktop().browse(e.getURL().toURI());
+                } catch (IOException | URISyntaxException e1) {
+                    // swallow exceptions
+                }
+            }
+        });
+        authExecutor = ApplicationManager.getApplication()
+            .executeOnPooledThread(() -> pullAuthenticationResult(ctx, deviceCode, callBack));
+        init();
+    }
+
+    private void pullAuthenticationResult(final AuthenticationContext ctx, final DeviceCode deviceCode,
+                                          final AuthenticationCallback<AuthenticationResult> callback) {
+        final long interval = deviceCode.getInterval();
+        long remaining = deviceCode.getExpiresIn();
+        while (remaining > 0 && authenticationResult == null) {
+            try {
+                remaining -= interval;
+                Thread.sleep(interval * 1000);
```

Comment:
Just an option, have you tried `RxJava` for async threads and retry logics, which would be easier to do that. 😄